Progress Report April 2021 (#272)
yuzu-emu/yuzu-emu.github.io: site/content/entry/yuzu-progress-report-apr-2021/index.md

+++
date = "2021-05-09T12:00:00-03:00"
title = "Progress Report April 2021"
author = "GoldenX86"
coauthor = "Honghoa"
forum = 0
+++ 

> Howdy yuz-ers! Another month gone, another progress report showing up. From new firmware versions to Vulkan fixes, we have quite a bit to talk about, so let’s get right into it!

Author: HP reference with the gone.

You just can’t beat WASD movement, especially when combined with mouse aiming.

{{< imgs
	"./keys.png| Rip and tear, press and click until it is done"
  >}}

Gamecube controllers connected to an adapter sometimes reported the joystick off-center, due to invalid data being sent during plugging.
[german77](https://github.com/german77) solved this [by taking 4 measurements during initialization](https://github.com/yuzu-emu/yuzu/pull/6243), ensuring that the controller finds its center, like any warrior should.

## User interface changes

[Kewlan](https://github.com/Kewlan) is back with more input improvements!

Starting with [adding a favourite option in the context menu,](https://github.com/yuzu-emu/yuzu/pull/6198) so you can force some of your games to the top of the list.

{{< imgs
	"./fav.png| Just right click them!"
  >}}

[Updating the results from a filtered search](https://github.com/yuzu-emu/yuzu/pull/6261) after you remove a directory. This fixes wrong behaviour if the user decided to do changes to the game folders while doing a filtered search.

> And lastly, fixing a [random bug that caused folders to expand or collapse when moving them in the game list.](https://github.com/yuzu-emu/yuzu/pull/6263)

Reviewer: ```suggestion
And lastly, they fixed a [random bug that caused folders to expand or collapse when moving them in the game list.](https://github.com/yuzu-emu/yuzu/pull/6263)
```

Several users reported that `Super Smash Bros. Ultimate` may get stuck when loading our web applet, making the whole game softlock.
Turns out the URL used by the applet was getting deleted before being used.
[Morph](https://github.com/Morph1984) got around this bug [by extending the lifetime of the URL.](https://github.com/yuzu-emu/yuzu/pull/6257)

## It’s 12 o’clock and it’s time for a new firmware!

> A big challenge made its entrance last month: the Nintendo Switch's firmware version was finally updated to Version 12.x, which has kept our developers quite busy figuring out what has changed and what needs to be implemented.

Reviewer: ```suggestion
A big challenge made its entrance last month: the Nintendo Switch's firmware version was finally updated to Version 12.X, which has kept our developers quite busy figuring out what has changed and what needs to be implemented.
```

> Thanks to joint efforts of [german77](https://github.com/german77), [epicboy](https://github.com/ameerj), [lioncash](https://github.com/lioncash), and [Morph](https://github.com/Morph1984), yuzu was able to update many system call ID tables ([PR #6153](https://github.com/yuzu-emu/yuzu/pull/6153), [PR #6154](https://github.com/yuzu-emu/yuzu/pull/6154)), services and function tables ([PR #6157](https://github.com/yuzu-emu/yuzu/pull/6157), [PR #6158](https://github.com/yuzu-emu/yuzu/pull/6158), [PR #6159](https://github.com/yuzu-emu/yuzu/pull/6159), [PR #6160](https://github.com/yuzu-emu/yuzu/pull/6160), [PR #6161](https://github.com/yuzu-emu/yuzu/pull/6161), [PR #6164](https://github.com/yuzu-emu/yuzu/pull/6164), [PR #6171](https://github.com/yuzu-emu/yuzu/pull/6171)), as well some system values ([PR #6155](https://github.com/yuzu-emu/yuzu/pull/6155)).

Reviewer: ```suggestion
Thanks to joint efforts of [german77](https://github.com/german77), [epicboy](https://github.com/ameerj), [lioncash](https://github.com/lioncash), and [Morph](https://github.com/Morph1984), yuzu was able to update many system call ID tables ([PR #6153](https://github.com/yuzu-emu/yuzu/pull/6153), [PR #6154](https://github.com/yuzu-emu/yuzu/pull/6154)), services and function tables ([PR #6157](https://github.com/yuzu-emu/yuzu/pull/6157), [PR #6158](https://github.com/yuzu-emu/yuzu/pull/6158), [PR #6159](https://github.com/yuzu-emu/yuzu/pull/6159), [PR #6160](https://github.com/yuzu-emu/yuzu/pull/6160), [PR #6161](https://github.com/yuzu-emu/yuzu/pull/6161), [PR #6164](https://github.com/yuzu-emu/yuzu/pull/6164), [PR #6171](https://github.com/yuzu-emu/yuzu/pull/6171)), as well as some system values ([PR #6155](https://github.com/yuzu-emu/yuzu/pull/6155)).
```

One of the many functions of the kernel in an operating system is to manage and distribute these resources to the processes that request them  — and for this, it must know the availability and limits of these resources.
Whenever a process is created or deleted, the hardware calculates the amount of memory that is free to use and being used by checking and updating the variables that keep track of these resources.

Previously, yuzu created processes in such a way that each instance of a process had their own value for `resource limits`.
This was a work-around to ensure that, whenever a process queried the kernel for the amount of memory available, a correct value was always returned.
This, however, is not how the hardware works.
Memory is a resource that is shared system-wide among all the processes, and the task of the kernel to coordinate how these resources are shared among the processes in the system.

With this PR, processes don't "own" their own instance of `resource limits`, and they now use the global kernel instance instead.

> [Dynarmic](https://github.com/MerryMage/dynarmic) — yuzu's JIT, used to translate the Switch’s ARM CPU instructions — was recently updated again, in order to [increase the size of the code cache](https://github.com/yuzu-emu/yuzu/pull/6132) (where yuzu stores the translated code), and add support for some unimplemented instructions from the `thumb32` set, expanding the amount of code that can be translated.

Reviewer: ```suggestion
[Dynarmic](https://github.com/MerryMage/dynarmic) — yuzu's JIT, used to translate the Switch’s ARM CPU instructions — was recently updated again in order to [increase the size of the code cache](https://github.com/yuzu-emu/yuzu/pull/6132) (where yuzu stores the translated code), and add support for some unimplemented instructions from the `thumb32` set, expanding the amount of code that can be translated.
```

[This PR](https://github.com/yuzu-emu/yuzu/pull/6214) allowing `Super Kirby Clash` and `Yo-Kai Watch 4` to boot.

{{< imgs
	"./kirby.png| Poyo!"
  >}}

[Morph](https://github.com/Morph1984) [stubbing `SetRequestExitToLibraryAppletAtExecuteNextProgramEnabled`](https://github.com/yuzu-emu/yuzu/pull/6234) and [german77](https://github.com/german77) [adding the `ectx:aw` service](https://github.com/yuzu-emu/yuzu/pull/6235) made `Pixel Game Maker Series Werewolf Princess Kaguya` boot!

{{< single-title-imgs
    "Right picture is from a more recent PR that will be mentioned in the next progress report (Pixel Game Maker Series Werewolf Princess Kaguya)"
    "./kaguya.png"
    "./kaguya2.png"
  >}}

The game currently has rendering bugs especially affecting Nvidia hardware. AMD users will have a better experience in this particular game for now.

## Graphics improvements

For our ~~thermonuclear~~ laptop users and heavy multitaskers, the legendary [degasus](https://github.com/degasus) has a treat for you. 
By [avoiding spin loops](https://github.com/yuzu-emu/yuzu/pull/6162) in the video core, more idle time is achieved with no performance loss.

> This means the GPU can enter `sleep` state more often, resulting in lower temperatures/power consumption, better battery life, or allowing the scheduler to take on other tasks waiting in the background, improving system responsiveness.

Reviewer: ```suggestion
This means the GPU can enter `sleep` state more often, resulting in lower temperatures/power consumption, better battery life, and allows the scheduler to take on other tasks waiting in the background, improving system responsiveness.
```

No more burned thighs while on the pause menu, laptop users!

> [Joshua-Ashton](https://github.com/Joshua-Ashton), the Frog Master from [DXVK](https://github.com/doitsujin/dxvk), brought us a couple of Vulkan fixes and improvements.

Reviewer: ```suggestion
[Joshua-Ashton](https://github.com/Joshua-Ashton), the Frog Master from [DXVK](https://github.com/doitsujin/dxvk), brought us a few Vulkan fixes and improvements.
```

First on the list is avoiding a device loss (GPU locks up and the driver restarts it) by [checking the return value of vkAcquireNextImageKHR.](https://github.com/yuzu-emu/yuzu/pull/6180)

Joshua also [enabled the individual features](https://github.com/yuzu-emu/yuzu/pull/6181) from `VK_EXT_robustness2`, instead of just enabling the extension.

And finally, [fixing a wrong offset for null vertex buffers](https://github.com/yuzu-emu/yuzu/pull/6182) to properly follow the Vulkan specification. 

> You never know when a driver will behave in erratic ways. Fixes like this allow developers to report bugs to the vendor’s driver development teams, thanks to confirming everything is within specs. Any resulting error falls under responsibility of the driver teams.

Reviewer: ```suggestion
You never know when a driver will behave in erratic ways. Fixes like this allow developers to report bugs to the vendors' driver development teams, thanks to confirming everything is within specs. Any resulting error falls under responsibility of the driver teams.
```

> Not only OpenGL had BGR issues, it turns out Vulkan can also suffer from blue madness when a game uses the `A1B5G5R5_UNORM` texture format.

Reviewer: ```suggestion
OpenGL wasn't the only API to have BGR issues, it turns out Vulkan can also suffer from "blue madness" when a game uses the `A1B5G5R5_UNORM` texture format.
```

[epicboy](https://github.com/ameerj) [swapped the Red and Blue channels](https://github.com/yuzu-emu/yuzu/pull/6238) of this format, fixing colours in games like `Pokémon Let’s Go Eevee/Pikachu`, `Shantae: Risky’s Revenge`, and many others!
You can see the result in the following pictures, before (left) and after (right).

{{< single-title-imgs
    "I didn’t want Pokémon Blue! (Pokémon Let's Go, Eevee!)"
    "./pokebug.png"
    "./pokefix.png"
  >}}

{{< single-title-imgs
    "That’s definitely not Vivi (Shantae: Risky’s Revenge)"
    "./shanbug.png"
    "./shanfix.png"
  >}}

## Linux Build System

[toastUnlimited](https://github.com/lat9nq) has been periodically improving yuzu’s build system for our Linux user base. 

> Not all our users want or are able to use the official AppImage, so not only did our build instructions needed to improve, but it was also equally important to refine the environment necessary to get yuzu running — namely, the dependencies, build process, etc.

Reviewer: ```suggestion
Not all our users want or are able to use the official AppImage, so not only did our build instructions need to be improved, but it was also equally important to refine the environment necessary to get yuzu running — namely, the dependencies, build process, etc.
```

The first step was to [make SDL2](https://github.com/yuzu-emu/yuzu/pull/6204) [an external dependency.](https://github.com/yuzu-emu/yuzu/pull/6207)

> This would allow us to ship the latest version without having to face Linux distributions incompatibilities due to different build flags (different characteristics being enabled), or outright outdated versions that aren’t able to support all of the required features that our input asks for.

Reviewer: ```suggestion
This would allow us to ship the latest version without having to face Linux distributions' incompatibilities due to different build flags (different characteristics being enabled), or outright outdated versions that aren’t able to support all of the required features that our input asks for.
```

> The changes to SDL2 introduced an incompatibility with the Conan-provided Boost version. [Downloading it as an external](https://github.com/yuzu-emu/yuzu/pull/6222) ensures compatibility, as most of the time the local packages provided by the distribution won’t be enough.

Reviewer: ```suggestion
The changes to SDL2 introduced an incompatibility with the Conan-provided Boost version. [Downloading it as an external dependency](https://github.com/yuzu-emu/yuzu/pull/6222) ensures compatibility, as most of the time the local packages provided by the distribution won’t be enough.
```

This change also makes our external download script cross-platform with Windows and other distributions that used to require special workarounds.

As a special bonus, due to all of these changes, Red Hat Enterprise Linux 8 and its derivatives can build yuzu in a much more user-friendly way.

For anyone interested, we keep updated build instructions [on this page.](https://github.com/yuzu-emu/yuzu/wiki/Building-For-Linux)

## Future projects

> This section has lately turned into Project Hades teasing, We won’t apologise for that. Here’s more info/teasing:

Reviewer: ```suggestion
This section has lately turned into Project Hades teasing, but we won’t apologize for that! Here’s more info/teasing:
```
